### ISSUE TITLE
[JENKINS-4071] jnlp slaves doesn't support HTTP Auth

### ISSUE THREAD

**Issue by timja** [timja/jenkins-gh-issues-poc-06-20#84]
<p>the jnlp slaves can't ship the auth credentials right now to get past a HTTP<br/>
Basic Authed hudson setup. This little patch may be a bit hacky and only support<br/>
Basic Auth, but it does the job for me.</p>

<p>$ java -jar slave.jar -noCertificateCheck -jnlpUrl<br/>
<a href="https://example.org/hudson/computer/NAME/slave-agent.jnlp" class="external-link" target="_blank" rel="nofollow noopener">https://example.org/hudson/computer/NAME/slave-agent.jnlp</a> -jnlpCredentials<br/>
user:password</p>

<p>Index: main/remoting/src/main/java/hudson/remoting/Launcher.java<br/>
===================================================================<br/>
&#8212; main/remoting/src/main/java/hudson/remoting/Launcher.java   (revision 19965)<br/>
+++ main/remoting/src/main/java/hudson/remoting/Launcher.java   (working copy)<br/>
@@ -94,6 +94,9 @@<br/>
     "Connection parameters are obtained by parsing the JNLP file.")<br/>
     public URL slaveJnlpURL = null;</p>

<p>+    @Option(name="-jnlpCredentials")<br/>
+    public String slaveJnlpCredentials = null;<br/>
+<br/>
     @Option(name="-cp",aliases="-classpath",metaVar="PATH",<br/>
     usage="add the given classpath elements to the system classloader.")<br/>
     public void addClasspath(String pathList) throws Exception {<br/>
@@ -185,6 +188,12 @@<br/>
 while (true) {<br/>
     try {<br/>
 URLConnection con = slaveJnlpURL.openConnection();<br/>
+if (con instanceof HttpURLConnection && slaveJnlpCredentials !=<br/>
null) </p>
{
+    HttpURLConnection http = (HttpURLConnection) con;
+   String userPassword = slaveJnlpCredentials;
+   String encoding = new sun.misc.BASE64Encoder().encode
(userPassword.getBytes());
+   http.setRequestProperty ("Authorization", "Basic " + encoding);
+       }
<p> con.connect();</p>

<p> if (con instanceof HttpURLConnection) {<br/>
@@ -216,6 +225,10 @@<br/>
 List<String> jnlpArgs = new ArrayList<String>();<br/>
 for( int i=0; i<argElements.getLength(); i++ )<br/>
 jnlpArgs.add(argElements.item<img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/information.png" height="16" width="16" align="absmiddle" alt="" border="0"/>.getTextContent());<br/>
+if (slaveJnlpCredentials != null) </p>
{
+       jnlpArgs.add("-credentials");
+      jnlpArgs.add(slaveJnlpCredentials);
+       }
<p> // force a headless mode<br/>
 jnlpArgs.add("-headless");<br/>
 return jnlpArgs;<br/>
Index: main/remoting/src/main/java/hudson/remoting/Engine.java<br/>
===================================================================<br/>
&#8212; main/remoting/src/main/java/hudson/remoting/Engine.java     (revision 19965)<br/>
+++ main/remoting/src/main/java/hudson/remoting/Engine.java     (working copy)<br/>
@@ -78,6 +78,7 @@</p>

<p>     private final String secretKey;<br/>
     public final String slaveName;<br/>
+    private String credentials;</p>

<p>     /**</p>
<ul>
	<li>See Main#tunnel in the jnlp-agent module for the details.<br/>
@@ -103,6 +104,10 @@<br/>
 this.tunnel = tunnel;<br/>
     }</li>
</ul>

<p>+    public void setCredentials(String creds) </p>
{
+this.credentials = creds;
+    }
<p>+<br/>
     public void setNoReconnect(boolean noReconnect) </p>
{
 this.noReconnect = noReconnect;
     }
<p>@@ -130,6 +135,11 @@</p>

<p>     // find out the TCP port<br/>
     HttpURLConnection con =<br/>
(HttpURLConnection)salURL.openConnection();<br/>
+if (con instanceof HttpURLConnection && credentials != null) </p>
{
+    HttpURLConnection http = (HttpURLConnection) con;
+   String encoding = new sun.misc.BASE64Encoder().encode
(credentials.getBytes());
+   http.setRequestProperty ("Authorization", "Basic " + encoding);
+       }
<p>     try </p>
{
 con.connect();
     }
<p> catch (IOException x) {<br/>
Index: main/remoting/src/main/java/hudson/remoting/jnlp/Main.java<br/>
===================================================================<br/>
&#8212; main/remoting/src/main/java/hudson/remoting/jnlp/Main.java  (revision 19965)<br/>
+++ main/remoting/src/main/java/hudson/remoting/jnlp/Main.java  (working copy)<br/>
@@ -64,6 +64,10 @@<br/>
     usage="Specify the Hudson root URLs to connect to.")<br/>
     public final List<URL> urls = new ArrayList<URL>();</p>

<p>+    @Option(name="-credentials",metaVar="USER:PASSWORD",<br/>
+    usage="Specify the Hudson credentials")<br/>
+    public String credentials;<br/>
+<br/>
     @Option(name="-noreconnect",<br/>
     usage="If the connection ends, don't retry and just exit.")<br/>
     public boolean noReconnect = false;<br/>
@@ -111,6 +115,8 @@<br/>
 urls, args.get(0), args.get(1));<br/>
 if(tunnel!=null)<br/>
     engine.setTunnel(tunnel);<br/>
+if(credentials!=null)<br/>
+    engine.setCredentials(credentials);<br/>
 engine.setNoReconnect(noReconnect);<br/>
 engine.start();<br/>
 engine.join();</p>

---
<details><summary><i>Originally reported by <a title="weigon" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=weigon">weigon</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-4071" target="_blank">jnlp slaves doesn't support HTTP Auth</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2009-10-23T14:55:09+00:00
<li><b>imported</b>: 2022-06-20
</ul></i>
</details>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<ul>
	<li>
	<ul>
		<li>
		<ul>
			<li>Issue 4072 has been marked as a duplicate of this issue. ***</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>:</i>
<p>Please put the patch as an attachment to avoid automatic line wraps.</p>

<p>There's one thing I don't understand, though &#8212; Even when the security is<br/>
enabled, all the URLs that JNLP slaves need to access is outside access control.</p>

<p>Are you setting up BASIC auth at the front end Apache or something?</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=weigon">weigon</a>:</i>
<p>Created an attachment (id=789)<br/>
added Basic Auth to JNPL</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=weigon">weigon</a>:</i>
<p>Yes, we use apache with auth in the frontend and hudson behind it. We have a<br/>
more than one service behind this frontend apache and auth only once to use them<br/>
all.</p>

<p>The auth-code of this patch is a bit hacky and only does Basic Auth.</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in hudson<br/>
User: : kohsuke<br/>
Path:<br/>
 trunk/hudson/main/remoting/src/main/java/hudson/remoting/Engine.java<br/>
 trunk/hudson/main/remoting/src/main/java/hudson/remoting/Launcher.java<br/>
 trunk/hudson/main/remoting/src/main/java/hudson/remoting/jnlp/Main.java<br/>
 trunk/www/changelog.html<br/>
<a href="http://fisheye4.cenqua.com/changelog/hudson/?cs=23113" class="external-link" target="_blank" rel="nofollow noopener">http://fisheye4.cenqua.com/changelog/hudson/?cs=23113</a><br/>
Log:<br/>
<span class="error">&#91;FIXED JENKINS-4071&#93;</span> Applied the change toward 1.331.</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in hudson<br/>
User: : kohsuke<br/>
Path:<br/>
 trunk/hudson/main/remoting/src/main/java/hudson/remoting/Engine.java<br/>
 trunk/www/changelog.html<br/>
<a href="http://jenkins-ci.org/commit/31576" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/31576</a><br/>
Log:<br/>
<span class="error">&#91;FIXED JENKINS-6262&#93;</span> HTTP auth support was added in <a href="https://issues.jenkins.io/browse/JENKINS-4071" title="jnlp slaves doesn&#39;t support HTTP Auth" class="issue-link" data-issue-key="JENKINS-4071"><del>JENKINS-4071</del></a>, so I've only integrated the portion that specifies the timeout and a call to disconnect(). Will be in 1.361.</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cforce101">cforce101</a>:</i>
<p>Is it possible to modify the jnlp this way that the user will be asked if he wann accept the self-signed certificate instead of using parameter "noCertificateCheck" on cmd line call?<br/>
Else is there any other solution to make the default jnlp  work with self-signed certificates when using ssl?</p>

**Comment by timja** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ringerc">ringerc</a>:</i>
<p>This doesn't change the example commands to show the authentication argument on the node info page, so you have to research how to get the slave to work once you hit this problem.</p>

**Comment by timja** [comment]
<i>[Originally duplicated by: <a href="https://github.com/timja/jenkins-gh-issues-poc-06-20/issues?q=in%3Atitle%20JENKINS-3734">JENKINS-3734</a>]</i>

**Comment by timja** [comment]
<i>[Originally duplicated by: <a href="https://github.com/timja/jenkins-gh-issues-poc-06-20/issues?q=in%3Atitle%20JENKINS-2774">JENKINS-2774</a>]</i>

**Comment by timja** [comment]
<i>[Originally duplicated by: <a href="https://github.com/timja/jenkins-gh-issues-poc-06-20/issues?q=in%3Atitle%20JENKINS-4072">JENKINS-4072</a>]</i>
